dashboard/modules/reporter/profile_manager.py

```diff
@@ -131,6 +143,115 @@ async def cpu_profile(
         )
         stdout, stderr = await process.communicate()
         if process.returncode != 0:
-            return False, _format_failed_pyspy_command(cmd, stdout, stderr)
+            return False, _format_failed_profiler_command(
+                cmd, self.profiler, stdout, stderr
+            )
         else:
             return True, open(profile_file_path, "rb").read()
+
+
+class MemoryProfilingManager:
+    def __init__(self, profile_dir_path: str):
+        self.profile_dir_path = Path(profile_dir_path)
+        self.profile_dir_path.mkdir(exist_ok=True)
+        self.profiler = "memray"
+
+    async def get_profile_result(
+        self, pid: int, format="flamegraph", leaks=False
+    ) -> (bool, str):
+        memray = shutil.which(self.profiler)
+        if memray is None:
+            return False, "memray is not installed"
+
+        profile_file_path = self.profile_dir_path / f"{pid}_memory_profiling.bin"
+        if not Path(profile_file_path).is_file():
+            return False, f"process {pid} has not been profiled"
+
+        profile_visualize_path = self.profile_dir_path / f"{pid}_memory_profiling.html"
+        if format == "flamegraph":
+            visualize_cmd = [
+                memray,
+                "flamegraph",
+                "-o",
+                profile_visualize_path,
+                "-f",
+            ]
+        elif format == "table":
+            visualize_cmd = [
+                memray,
+                "table",
+                "-o",
+                profile_visualize_path,
+                "-f",
```

Comment:
Same recommendation as above– we should avoid overwriting the tables if possible.